syncer/receiver/downloader.go

```diff
@@ -102,22 +105,46 @@ func (d *Downloader) LoadOnce(ctx context.Context, ni snapshot.NameInfo) error {
 	d.r.storageLoadHealth.AddSuccess()
 
 	metricSnapshotsLoadBytes.Add(float64(len(data)))
+
+	// Limit number of decompressed snapshots in memory
+	// CAUTION: we cannot defer the Release, check all error paths!
+	token := d.r.decompressedSnapshotLimit.Acquire()
+
 	t1 := time.Now()
 
 	msg, err := snapshot.LoadData(data)
 	if err != nil {
+		d.l.Debug("Returning DecompressedSnapshotToken")
+		token.Release()
 		// This snapshot is considered corrupt, we will ignore it from now on
 		d.r.MarkCorrupt(ni.FullName, err)
 		d.last = ni
 		return err
 	}
 
+	// Release the download token once we have released the downloaded snapshot
+	data = nil // allow it to be freed
+	_ = data   // silence linter
+	downloadToken.Release()
+
 	// Make snapshot available to the syncer, replacing any previous one
 	// that has not been loaded yet.
 	d.r.mu.Lock()
+	// FIXME: use *snapshot.Update pointer in APIs with new tokens
 	d.r.snapshotsByInstance[d.instance] = snapshot.Update{
 		Snapshot: msg,
 		NameInfo: ni,
+		OnClose: func(u *snapshot.Update) {
+			if u.Snapshot == nil {
+				return // already called?
```

Comment:
There are two ways we can handle calling Close() twice:

1. Ignore a second call.
2. panic()

The advantage of the first is that you can always `defer u.Close()`, even if you explicitly call it somewhere. Having this option may reduce the chance that you forget to call it in the first place.

The second one us strict about you calling it only once. This may detect incorrect logic in some cases, but actually increases the chance that you forget to call it in some code paths.

In this case I picked option 1, because it seemed like the least bad option.

The bigger issue happens when Close is _not_ called. Perhaps we should add a `runtime.SetFinalizer` to detect that scenario, and then panic?